### ISSUE TITLE
Allow custom property serialization

### ISSUE THREAD

**Issue by myronmarston** [riak-ripple/ripple#143]
For one of my ripple models, I have a property (`dates_scheduled`) that I want to be a `Set`.  This has a few advantages over an array:
- A set ensures uniqueness.  When you do `my_set << object_already_in_set`, it's a no op.
- `Set#include?` is a constant time operation; `Array#include?` is linear time.

I hacked something together in my model:

```
require 'set'

class Subscription
  include Ripple::Document

  def dates_scheduled
    @dates_scheduled ||= Set.new
  end

  def dates_scheduled=(val)
    @dates_scheduled = Set.new(val.map { |v| Date.iso8601(v) })
  end

  private

    def attributes_for_persistence
      super.tap do |attrs|
        attrs['dates_scheduled'] = dates_scheduled.to_a
      end
    end
end
```

This works decently well, but has a couple flaws:
- It relies upon `attribues_for_persistence`, which, as far as I can tell, is not a documented, public API.
- `dates_scheduled` is not a true property.  It doesn't get all the other property behavior (dirty tracking, `dates_scheduled?` predicate, etc.)

It would be nice if ripple supported this directly, but I'm not sure how it should work.  One simple idea is to support an API like this:

```
class JSONSerializableSet < Set
  def to_json
    to_a.to_json
  end

  def self.from_json(json)
    array = YAML.load(json)
    new(array)
  end
end

class Subscription
  include Ripple::Document
  property :dates_scheduled, JSONSerializableSet
end
```

Basically, users could provide a custom type, as long as the type defines `#to_json` and `.from_json`.  However, I'm not sure how this would be implemented--it looks to me like the JSON deserialization happens in `RObject#deserialize` (defined in riak-client), and that simply uses `YAML.load` on the entire thing.  I don't know how we'd hook a user's custom type in there.

I'm happy to work on this feature once we get a consensus about how it should work (assuming Sean's on board with the feature).
